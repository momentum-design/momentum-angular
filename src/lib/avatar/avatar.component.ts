/** @component avatar */

import {
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnChanges,
  OnInit,
  Output,
  ViewChild,
} from '@angular/core';

export type AvatarSize =
  | 'xsmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xlarge'
  | 18
  | 24
  | 28
  | 36
  | 40
  | 44
  | 52
  | 56
  | 72
  | 80
  | 84;
export type AvatarType =
  | ''
  | 'active'
  | 'bot'
  | 'call'
  | 'dnd'
  | 'group'
  | 'inactive'
  | 'meeting'
  | 'ooo'
  | 'presenting'
  | 'self'
  | 'typing';

@Component({
  selector: 'md-avatar',
  template: `
    <button
      md-button
      *ngIf="clickable; else avatarTpl"
      [attr.aria-label]="ariaLabel"
      [class]="buttonClassName"
      [circle]="true"
      [removeStyle]="true"
    >
      <ng-container *ngTemplateOutlet="avatarTpl"></ng-container>
    </button>

    <ng-template #avatarTpl>
      <div
        *ngIf="clickable; else overlayTpl"
        class="md-avatar"
        [ngClass]="[
          'md-avatar--clickable',
          (type && 'md-avatar--' + type) || '',
          (size && 'md-avatar--' + size) || '',
          (theme && 'md-avatar--' + theme) || '',
          (isDecrypting && 'md-decrypting') || '',
          className || ''
        ]"
        [title]="!hideDefaultTooltip ? title : ''"
      >
        <ng-container *ngTemplateOutlet="overlayTpl"></ng-container>
      </div>
    </ng-template>

    <ng-template #overlayTpl>
      <ng-container *ngTemplateOutlet="childrenTpl"></ng-container>
      <md-loading *ngIf="type === 'typing'"></md-loading>
      <span *ngIf="failureBadge" class="md-avatar__failure-badge"></span>
      <span
        *ngIf="hasNotification"
        class="md-avatar__notification-badge"
      ></span>
    </ng-template>

    <ng-template #childrenTpl>
      <span
        *ngIf="type === 'self'; else imageTpl"
        class="md-avatar__self"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
      >
        <md-icon
          [name]="
            size === 40 || size === 'medium'
              ? 'chat-active_18'
              : 'chat-active_16'
          "
          color="blue-50"
        ></md-icon>
      </span>
    </ng-template>

    <ng-template #imageTpl>
      <ng-container *ngIf="src && !isImageErrored; else iconTpl">
        <ng-container *ngIf="title && !isImageLoaded">
          <ng-container *ngTemplateOutlet="letterTpl"></ng-container>
        </ng-container>
        <img
          #image
          [alt]="alt"
          class="md-avatar__img"
          [ngClass]="{ 'md-avatar__img--hidden': !isImageLoaded }"
          draggable="false"
          (error)="handleImgError()"
          (load)="handleImgLoaded()"
          [src]="src"
        />
      </ng-container>
    </ng-template>

    <ng-template #iconTpl>
      <span
        *ngIf="icon; else letterTpl"
        class="md-avatar__icon"
        [ngClass]="{ 'md-avatar__icon--overview': isOverview }"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
      >
        <md-icon [name]="icon"></md-icon>
      </span>
    </ng-template>

    <ng-template #letterTpl>
      <span
        class="md-avatar__letter"
        [ngClass]="{
          'md-decrypting': isDecrypting,
          'md-avatar__img--hidden': isImageLoaded
        }"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
        >{{ getInitials() }}</span
      >
    </ng-template>
  `,
  host: {
    '[attr.title]': '!hideDefaultTooltip ? title : ""',
    class: 'md-avatar md-avatar--medium',
    '[class.md-decrypting]': '!clickable && isDecrypting',
  },
})
export class AvatarComponent implements OnInit, OnChanges {
  @Input() public ariaLabel: string = '';
  /** @option Image alt tag | '' */
  @Input() public alt: string = '';
  /** @option Set Avatar background color | '' */
  @Input() public backgroundColor: string = '';
  /** @option Optional css class string for button | '' */
  @Input() public buttonClassName: string = '';
  /** @option Optional css class string for Avatar component | null */
  @Input() public className: string;
  /** @option Set Avatar text color | '' */
  @Input() public color: string = '';
  /** @option Set existance of a failureBadge on the Avatar | false */
  @Input() public failureBadge: boolean = false;
  /** @option Set existance of a notification on the Avatar | false */
  @Input() public hasNotification: boolean = false;
  /** @option Set the visibility of Avatar's default tooltip | false */
  @Input() public hideDefaultTooltip: boolean = false;
  /** @option Optional icon name for the Avatar | null */
  @Input() public icon: string;
  /** @option Set if Avatar's content is decrypting | false */
  @Input() public isDecrypting: boolean = false;
  /** @option Set existance of Avatar's Overview | false */
  @Input() public isOverview: boolean = false;
  /** @option Set the size of the Avatar from one of the preconfigured options | 'medium' */
  @Input()
  get size(): AvatarSize {
    return this._size;
  }
  set size(value: AvatarSize) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._size}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._size = value;
  }
  /** @option Optional image source for the Avatar | null */
  @Input() public src: string;
  /** @option Optional Avatar color theme | null */
  @Input()
  get theme(): string {
    return this._theme;
  }
  set theme(value: string) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._theme}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._theme = value;
  }
  /** @option set Avatar title / user's name | null */
  @Input() public title: string;
  /** @option optional Avatar type | '' */
  @Input()
  get type(): AvatarType {
    return this._type;
  }
  set type(value: AvatarType) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._type}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._type = value;
  }

  @Output() click: EventEmitter<any> = new EventEmitter();

  @ViewChild('image') image: ElementRef;

  private _size: AvatarSize = 'medium';
  private _type: AvatarType = '';
  private _theme: string;
  public clickable: boolean = false;
  public isImageLoaded: boolean = false;
  public isImageErrored: boolean = false;

  constructor(private elementRef: ElementRef) {}

  ngOnInit() {
    const img = this.image;
    if (img && img.nativeElement.complete) {
      this.handleImgLoaded();
    }
    this.clickable = this.click.observers.length > 0;
  }

  ngOnChanges(changes: any) {
    if (changes.src && changes.src.previousValue !== changes.src.currentValue) {
      this.handleImgChange();
    }
  }

  handleImgChange() {
    this.isImageLoaded = false;
    this.isImageErrored = false;
  }

  handleImgError() {
    this.isImageErrored = true;
  }

  handleImgLoaded() {
    this.isImageLoaded = true;
  }

  getInitials() {
    if (!this.title || !this.title.replace(/\s/g, '').length) {
      return '';
    }
    const letters = [];
    const words = this.title.trim().split(/ +/);
    const repeatTimes = Math.min(
      (this.type === 'group' && 1) || 2,
      words.length
    );
    for (let i = 0; i < repeatTimes; i++) {
      letters.push(String.fromCodePoint(words[i].codePointAt(0)));
    }
    return letters.join('');
  }
}
