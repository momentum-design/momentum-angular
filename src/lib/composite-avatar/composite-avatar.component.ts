import { Component, HostBinding, Input } from '@angular/core';

export type CompositeAvatarType =
  | 'small'
  | 'medium'
  | 'large'
  | 28
  | 40
  | 84
  | 135;

@Component({
  selector: 'md-composite-avatar',
  template: `
    <ng-content></ng-content>
  `,
})
export class CompositeAvatarComponent {
  @Input() private className: string = '';
  @Input() private size: CompositeAvatarType = 'medium';

  @HostBinding('class') get _class(): string {
    return (
      'md-composite-avatar' +
      `${(this.size && ` md-composite-avatar--${this.size}`) || ''}` +
      `${(this.className && ` ${this.className}`) || ''}`
    );
  }

  constructor() {}
}
