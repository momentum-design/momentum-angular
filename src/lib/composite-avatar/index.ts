export * from './composite-avatar.module';
export * from './composite-avatar.component';
