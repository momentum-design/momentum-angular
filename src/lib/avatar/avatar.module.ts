import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from '../button';
import { IconModule } from '../icon';
import { LoadingModule } from '../loading';
import { AvatarComponent } from './avatar.component';

@NgModule({
  imports: [CommonModule, ButtonModule, IconModule, LoadingModule],
  declarations: [AvatarComponent],
  exports: [AvatarComponent],
})
export class AvatarModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: AvatarModule,
      providers: [],
    };
  }
}
