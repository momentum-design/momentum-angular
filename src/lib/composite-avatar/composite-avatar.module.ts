import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AvatarModule } from '../avatar';
import { CompositeAvatarComponent } from './composite-avatar.component';

@NgModule({
  imports: [CommonModule, AvatarModule],
  declarations: [CompositeAvatarComponent],
  exports: [CompositeAvatarComponent],
})
export class CompositeAvatarModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CompositeAvatarModule,
      providers: [],
    };
  }
}
