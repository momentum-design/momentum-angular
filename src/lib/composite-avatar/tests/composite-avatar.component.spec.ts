import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { AvatarModule } from '../../avatar';
import { CompositeAvatarModule } from '../composite-avatar.module';

describe('CompositeAvatarComponent', () => {
  @Component({
    selector: 'test-app',
    template: `
      <md-composite-avatar id="test-composite-avatar">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="small" id="test-composite-avatar-small">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="28" id="test-composite-avatar-28">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="40" id="test-composite-avatar-40">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="84" id="test-composite-avatar-84">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="large" id="test-composite-avatar-large">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="135" id="test-composite-avatar-135">
        <md-avatar title="test1"></md-avatar>
        <md-avatar title="test2"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="large" id="test-composite-avatar-single">
        <md-avatar title="test1"></md-avatar>
      </md-composite-avatar>

      <md-composite-avatar size="large" id="test-composite-avatar-wrapped">
        <div>
          <md-avatar title="test1"></md-avatar>
        </div>
      </md-composite-avatar>
    `,
  })
  class TestAppComponent {}

  let fixture: ComponentFixture<TestAppComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestAppComponent],
      imports: [AvatarModule, CompositeAvatarModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestAppComponent);
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  describe('composite avatars of different sizes  ', () => {
    it('when size is small/28', () => {
      fixture.detectChanges();
      let element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-small'
      );
      expect(element.className).toContain('md-composite-avatar--small');
      element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-28'
      );
      expect(element.className).toContain('md-composite-avatar--28');
    });

    it('when size is medium/40', () => {
      fixture.detectChanges();
      let element = fixture.nativeElement.querySelector(
        '#test-composite-avatar'
      );
      expect(element.className).toContain('md-composite-avatar--medium');
      element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-40'
      );
      expect(element.className).toContain('md-composite-avatar--40');
    });

    it('when size is 84', () => {
      fixture.detectChanges();
      const element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-84'
      );
      expect(element.className).toContain('md-composite-avatar--84');
    });

    it('when size is large/135', () => {
      fixture.detectChanges();
      let element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-large'
      );
      expect(element.className).toContain('md-composite-avatar--large');
      element = fixture.nativeElement.querySelector(
        '#test-composite-avatar-135'
      );
      expect(element.className).toContain('md-composite-avatar--135');
    });
  });

  it('should not throw an error when 2 Avatar components are not present as children', () => {
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector(
      '#test-composite-avatar-wrapped'
    );
    expect(element.className).toContain('md-composite-avatar--large');
  });
});
