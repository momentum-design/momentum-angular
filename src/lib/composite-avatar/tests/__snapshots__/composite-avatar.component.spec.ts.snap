// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompositeAvatarComponent should match snapshot 1`] = `
<test-app>
  <md-composite-avatar
    class="md-composite-avatar md-composite-avatar--medium"
    id="test-composite-avatar"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--small"
    id="test-composite-avatar-small"
    ng-reflect-size="small"
    size="small"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--28"
    id="test-composite-avatar-28"
    ng-reflect-size="28"
    size="28"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--40"
    id="test-composite-avatar-40"
    ng-reflect-size="40"
    size="40"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--84"
    id="test-composite-avatar-84"
    ng-reflect-size="84"
    size="84"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--large"
    id="test-composite-avatar-large"
    ng-reflect-size="large"
    size="large"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--135"
    id="test-composite-avatar-135"
    ng-reflect-size="135"
    size="135"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test2"
      title="test2"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--large"
    id="test-composite-avatar-single"
    ng-reflect-size="large"
    size="large"
  >
    <md-avatar
      class="md-avatar md-avatar--medium"
      ng-reflect-title="test1"
      title="test1"
    >






      <span
        class="md-avatar__letter"
        ng-reflect-klass="md-avatar__letter"
        ng-reflect-ng-class="[object Object]"
        ng-reflect-ng-style="[object Object]"
      >
        t
      </span>














    </md-avatar>
  </md-composite-avatar><md-composite-avatar
    class="md-composite-avatar md-composite-avatar--large"
    id="test-composite-avatar-wrapped"
    ng-reflect-size="large"
    size="large"
  >
    <div>
      <md-avatar
        class="md-avatar md-avatar--medium"
        ng-reflect-title="test1"
        title="test1"
      >






        <span
          class="md-avatar__letter"
          ng-reflect-klass="md-avatar__letter"
          ng-reflect-ng-class="[object Object]"
          ng-reflect-ng-style="[object Object]"
        >
          t
        </span>














      </md-avatar>
    </div>
  </md-composite-avatar>
</test-app>
`;
